core/state/statedb.go

```diff
@@ -154,25 +160,30 @@ func newStateDB(root common.Hash, db Database, snaps *snapshot.Tree) (*StateDB,
 		journal:             newJournal(),
 		hasher:              crypto.NewKeccakState(),
 	}
-	tr, err := db.OpenTrie(root)
-	if err != nil {
-		return nil, err
-	}
-	sdb.trie = tr
 	if sdb.snaps != nil {
 		if sdb.snap = sdb.snaps.Snapshot(root); sdb.snap != nil {
 			sdb.snapDestructs = make(map[common.Address]struct{})
 			sdb.snapAccounts = make(map[common.Address][]byte)
 			sdb.snapStorage = make(map[common.Address]map[string][]byte)
 		}
 	}
+
+	snapVerified := sdb.snap != nil && sdb.snap.Verified()
+	tr, err := db.OpenTrie(root)
+	// return error when 1. failed to open trie and 2. the snap is not nil and done verification
+	if err != nil && (sdb.snap == nil || snapVerified) {
```

Comment:
according to the comment, should the condition here be `sdb.snap != nil && snapVerified`

Reply:
if `sdb.snap == nil`,  we won't allow open trie error.

Reply:
"err != nil" means OpenTrie failed, newStateDB should fail as well?
newStateDB will success now, if OpenTrie fail when snap!=nil and snap verify failed.
It is make sense?

Reply:
> "err != nil" means OpenTrie failed, newStateDB should fail as well? newStateDB will success now, if OpenTrie fail when snap!=nil and snap verify failed. It is make sense?
Got it, the behavior changed.
With pipeline commit, trie tree may not ready since IntermediateRoot() will be executed asynchronously.
OpenTrie is not necessary and a failed OpenTrie is probably not matter, since state verify is probably to success.